src/sql/src/plan/expr.rs

```diff
@@ -340,17 +340,19 @@ impl VisitChildren<HirScalarExpr> for WindowExpr {
 /// A window function with its parameters.
 ///
 /// There are three types of window functions:
-/// - scalar window functions, that return a different scalar value for each
-/// row within a partition that depends exclusively on the position of the row
-/// within the partition;
-/// - value window functions, that return a scalar value for each row within a
-/// partition that might be computed based on a single previous, current or
-/// following row;
-/// - aggregate window functions, that return a computed value for the row that
-/// depends on multiple other rows within the same partition. Aggregate window
-/// functions can be in some cases be computed by joining the input relation
-/// with a reduction over the same relation that computes the aggregation using
-/// the partition key as its grouping key.
+/// - scalar window functions, which return a different scalar value for each
+///   row within a partition that depends exclusively on the position of the row
+///   within the partition;
```

Comment:
> My understanding is that using ROW_NUMBER() requires us to do just as much work as using LAG() etc.

This is correct.

> If it's a simple scalar expression how does it know where it is in the window?

Don't get too hung up on the term "scalar window function"; this term seems to be actually a misnomer. I always assumed that this is a real term, but now I googled it, and it turns out this is not a thing at all. The standard name seems to be "ranking window function" for this type of window functions, see e.g.
- https://medium.com/@ritusantra/window-functions-in-ms-sql-server-0221c7656fec
- https://medium.com/@ritusantra/window-functions-in-ms-sql-server-0221c7656fec
- https://www.datacamp.com/cheat-sheet/sql-window-functions-cheat-sheet

I'm trying to think about why the term "scalar window function" was introduced in our code, but it doesn't really make any sense to me, since all the other window functions also yield a scalar as a result.

I think I'll rename this in our code in a follow-up PR.